### PR TITLE
PERF-#6876: Skip the masking stage on 'iloc' where beneficial

### LINKED ISSUE #6876: df.take is much slower against pandas
On a machine with 192 CPUs.

```python
# import pandas as pd
import modin.pandas as pd
import numpy as np
import time

df = pd.DataFrame(data=np.random.randint(99999, 99999999, size=(100000000,4)),
                columns=['C1','C2','C3','C4'])

to_take = np.random.randint(0, 100000000, size=80000000)
t0 = time.time()
df.take(to_take, axis=0)
t1 = time.time()
print('time to take: ', t1 - t0)
# time for take:  44.71011233329773 in Modin
# time for take:  3.082368850708008 in pandas
```

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

This PR extends the idea originally introduced in #6423.

<b> What's the idea </b>

The masking pipeline consists of two stages:
1. Compute individual indexers for each partition and mask them.
2. If the indexing forces reordering, apply a full-column reordering function to the result of the first stage.

In theory, the second stage can function without the first one, the first stage is needed in order to reduce the dimension of the input for the second stage. However, sometimes the first stage can be quite expensive, and it could be reasonable to skip it and jump to the second one right away.

This PR extends the number of cases where we skip the first stage with the cases where we know for sure that the reordering stage will be required. Based on the measurements below, I found out the following condition that controls this:

```python
# 'base_num_cols' specifies the number of columns that the dataframe should have
# in order to jump to 'reordered_labels' in case of len(row_positions) / len(self) >= base_ratio;
# these variables may be a subject to change in order to tune performance more accurately
base_num_cols = 10
base_ratio = 0.2
# Example:
#   len(self.columns): 10 == base_num_cols -> min ratio to jump to reorder_labels: 0.2 == base_ratio
#   len(self.columns): 15 -> min ratio to jump to reorder_labels: 0.3
#   len(self.columns): 20 -> min ratio to jump to reorder_labels: 0.4
#   ...
#   len(self.columns): 49 -> min ratio to jump to reorder_labels: 0.98
#   len(self.columns): 50 -> min ratio to jump to reorder_labels: 1.0
#   len(self.columns): 55 -> min ratio to jump to reorder_labels: 1.0
#   ...
if rows_reordering_required and len(row_positions) * base_num_cols >= min(
    all_rows * len(self.columns) * base_ratio, len(row_positions) * base_num_cols
):
    # jump to reorder_labels
```

<details><summary>measurements with MODIN_CPUS=44</summary>

image

</details>

The reproducer from the issue (#6876) now performs like this:

image

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #6876 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -1161,12 +1161,22 @@ def _take_2d_positional(
...
+                    indexer = np.array(indexer, dtype=np.int64)
             indexers.append(indexer)
         row_positions, col_positions = indexers
 
         if col_positions is None and row_positions is None:
             return self.copy()
 
+        # quite fast check that allows skip sorting
+        must_sort_row_pos = row_positions is not None and not np.all(
+            row_positions[1:] >= row_positions[:-1]
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1174]: this check was copied from [`_maybe_reorder_labels`](https://github.com/modin-project/modin/blob/097ea527c8e3f099e1f252b067a1d5eb055ad0b5/modin/core/dataframe/pandas/dataframe/dataframe.py#L1294) and it indeed works pretty fast:
```python
import numpy as np
from timeit import default_timer as timer

arr = np.random.randint(0, 100_000_000, size=100_000_000)

t1 = timer()
must_sort_col_pos = np.all(arr[1:] >= arr[:-1])
print(timer() - t1) # 0.09s
```

**Comment by dchigarev** [comment]
@anmyachev I believe you're the most suitable person to review this

**Comment by anmyachev** [comment]
@dchigarev The picture won't open with "measurements with MODIN_CPUS=44".

**Comment by YarShev** [comment]
> @dchigarev The picture won't open (to full size) with "measurements with MODIN_CPUS=44".

works to me

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 2]
```diff
@@ -1175,18 +1185,40 @@ def _take_2d_positional(
             all_rows = None
             if self.has_materialized_index:
                 all_rows = len(self.index)
-            elif self._row_lengths_cache:
+            elif self._row_lengths_cache or must_sort_row_pos:
                 all_rows = sum(self._row_lengths_cache)
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1189]: Should we use `row_lengths` here?
```suggestion
                all_rows = sum(self.row_lengths)
```

**Comment by anmyachev** [comment]
> > @dchigarev The picture won't open (to full size) with "measurements with MODIN_CPUS=44".
> 
> works to me

Apparently I had a bug, now it also works for me.

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 3]
```diff
@@ -1175,18 +1185,40 @@ def _take_2d_positional(
...
+            #   ...
+            #   len(self.columns): 49 -> min ratio to jump to reorder_labels: 0.98
+            #   len(self.columns): 50 -> min ratio to jump to reorder_labels: 1.0
+            #   len(self.columns): 55 -> min ratio to jump to reorder_labels: 1.0
+            #   ...
+            if (all_rows and len(row_positions) > 0.9 * all_rows) or (
+                must_sort_row_pos
+                and len(row_positions) * base_num_cols
+                >= min(
+                    all_rows * len(self.columns) * base_ratio,
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1209]: Is it performance-wise safe to materialize columns here? (Are they already materializing somewhere nearby or not?)

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1209]: no, they're not materialized explicitly anywhere nearby, but this value is required to properly branch here, so I guess I have no good choices here

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1209]: my assumption is that columns are more likely to be pre-computed than indices, so accessing them shouldn't always trigger computations

Comment (anmyachev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1209]: Ok, we can try to use `self.column_widths` in the future if needed

**Review by anmyachev — APPROVED** [review]
LGTM!

**Pull request merged**